### PR TITLE
[RAPPS] Add error messages on all download failure cases (CORE-16851)

### PR DESCRIPTION
**Pull request by KyleKatarn31**
## Purpose

- Current RAPPS has many failure conditions during download process where process stops without user awareness nor logging, making root cause analysis difficult / impossible

JIRA issue: [CORE-16851](https://jira.reactos.org/browse/CORE-16851)

## Proposed changes

- Systematic error messages presented to the users when user awareness is required
- Additionnal logging when no additionnal user awareness is required (cancellation,...)
- All messages added to localization files

image

### CODE REVIEW

**Comment by KyleKatarn31** [comment]
@kernelbin : for review.

File: base/applications/rapps/loaddlg.cpp [hunk 1]
```diff
@@ -640,7 +651,10 @@ unsigned int WINAPI CDownloadManager::ThreadFunc(LPVOID param)
         }
 
         if (!hOpen)
+        {
+            MessageBox_LoadString(hMainWnd, IDS_UNABLE_TO_INTACCESS);
```
Comment (binarymaster) [inline base/applications/rapps/loaddlg.cpp:655]: You can use some API function that translates error code into localized text message, so you don't need to add all these messages in app.

Comment (KyleKatarn31) [reply base/applications/rapps/loaddlg.cpp:655]: Which API do you have in mind ? Most messages here are implementation-specific, not simply GetLastError(). Did I miss something ?

Comment (katahiromz) [reply base/applications/rapps/loaddlg.cpp:655]: FormatMessage with HMODULE.

Comment (KyleKatarn31) [reply base/applications/rapps/loaddlg.cpp:655]: Sorry but i'm missing the purpose.... what would be the benefit here ? 1 single IDS_UNABLE_TO like "Error occured : %d" ? Tailored error message looks more explicit for end user as cause may differ (configuration of path, disk capacity, proxy settings....)

Comment (binarymaster) [reply base/applications/rapps/loaddlg.cpp:655]: > 1 single IDS_UNABLE_TO like "Error occured : %d" ?

It's applicable to all error strings you added.

Comment (KyleKatarn31) [reply base/applications/rapps/loaddlg.cpp:655]: Sorry but I still don't undersand what you expect... In any case, i'll have the same number of strings to be translated and more code.... May I have a very simple PoC of what you suggest ?

**Comment by kernelbin** [comment]
I saw you used a lot of ```goto end;``` and something like this.
why don't use SEH instead?

**Comment by KyleKatarn31** [comment]
> 
> 
> I saw you used a lot of `goto end;` and something like this.
> why don't use SEH instead?

Purpose here is to improve user awareness upon failure, no full RAPPS download management rede-sign. That's more a topic reserved for GSoC work :)

**Comment by kernelbin** [comment]
> > I saw you used a lot of `goto end;` and something like this.
> > why don't use SEH instead?
> 
> Purpose here is to improve user awareness upon failure, no full RAPPS download management rede-sign. That's more a topic reserved for GSoC work :)

ok   : )

File: base/applications/rapps/loaddlg.cpp (hunk 1 above)
Comment (KyleKatarn31) [reply base/applications/rapps/loaddlg.cpp:655]: Same design solution as for https://github.com/reactos/reactos/pull/2653

File: base/applications/rapps/loaddlg.cpp [hunk 2]
```diff
@@ -508,6 +510,27 @@ VOID CDownloadManager::UpdateProgress(
     }
 }
 
+VOID ShowLastError(HWND hWndOwner,
+                        DWORD dwLastError)
```
Comment (binarymaster) [inline base/applications/rapps/loaddlg.cpp:514]:
```suggestion
VOID ShowLastError(
    HWND hWndOwner,
    DWORD dwLastError)
```
Formatting.

Comment (KyleKatarn31) [reply base/applications/rapps/loaddlg.cpp]: Fixed, thanks !

File: base/applications/rapps/loaddlg.cpp [hunk 3]
```diff
@@ -508,6 +510,28 @@ VOID CDownloadManager::UpdateProgress(
...
+                       dwLastError,
+                       LANG_USER_DEFAULT,
+                       (LPTSTR)&lpMsg,
+                       0, NULL))
+    {
+        return;
+    }
+
+    MessageBox(hWndOwner, lpMsg, NULL, MB_OK | MB_ICONERROR);
+    LocalFree(lpMsg);
```
Comment (kernelbin) [inline base/applications/rapps/loaddlg.cpp:532]: I suggest using Wide Char. ( FormatMessageW, MessageBoxW, .... )

Comment (KyleKatarn31) [reply base/applications/rapps/loaddlg.cpp:532]: Yes of course. I'll fix

**Comment by KyleKatarn31** [comment]
Same comment to be applied on : https://github.com/reactos/reactos/pull/2653/ (i'll use wide char too)

**Comment by KyleKatarn31** [comment]
All comments are now fully integrated.
Last review round before merge ?

**Comment by KyleKatarn31** [comment]
Following today's discussion in Dev channel, I will not modify this code.
Either you merge as is, or you close and discard.

**Comment by KyleKatarn31** [comment]
Thanks !

**Pull request merged**